### PR TITLE
CLOUDSTACK-3223 Exception observed while creating CPVM in VMware Setup with DVS

### PR DESCRIPTION
**Pull request by sateesh-chodapuneedi**
Jira
===
CLOUDSTACK-3223 Exception observed while creating CPVM in VMware Setup with DVS

Issue
=====
Exception observed while creating CPVM in VMware Setup with DVS.
Observed error `"StartCommand failed due to Exception: com.vmware.vim25.AlreadyExists."`
This is due to concurrent attempts to create same dv portgroup on same dvSwitch by
manager threads of CPVM and SSVM when both are started at same time.

Fix
===
Synchronize api calls to create/update dvportgroup.
Also maintaing local cache to avoid multiple fetch api calls to vCenter
when multiple threads try to create same object.

Signed-off-by: Sateesh Chodapuneedi <sateesh.chodapuneedi@accelerite.com>

### CODE REVIEW

**Comment by sateesh-chodapuneedi** [comment]
@borisstoyanov Can you please run trillian tests for this PR?

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test centos6 vmware-60u1

**Comment by borisstoyanov** [comment]
@blueorangutan test centos7 vmware-60u2

**Comment by sateesh-chodapuneedi** [comment]
@ping sureshanaparti @anshul1886

File: vmware-base/src/com/cloud/hypervisor/vmware/mo/DistributedVirtualSwitchMO.java [hunk 1]
```diff
@@ -36,24 +37,67 @@
...
+                ManagedObjectReference task = _context.getService().addDVPortgroupTask(_mor, dvPortGroupSpecArray);
+                if (!_context.getVimClient().waitForTask(task)) {
+                    throw new Exception("Failed to create dvPortGroup " + dvPortGroupSpec.getName());
+                } else {
+                    if (s_dvPortGroupCacheMap.containsKey(dvSwitchInstance)) {
+                        dvPortGroupList = s_dvPortGroupCacheMap.get(dvSwitchInstance);
+                        if (dvPortGroupList == null) {
+                            dvPortGroupList = new ArrayList<String>();
+                        }
+                        dvPortGroupList.add(dvPortGroupName); //does this update map?
```
Comment (sureshanaparti) [inline vmware-base/src/com/cloud/hypervisor/vmware/mo/DistributedVirtualSwitchMO.java]: @sateesh-chodapuneedi `s_dvPortGroupCacheMap` update required here?

Comment (sateesh-chodapuneedi) [inline vmware-base/src/com/cloud/hypervisor/vmware/mo/DistributedVirtualSwitchMO.java]: Yes @sureshanaparti, you are right the `dvPortGroupList` ref would take care of the update.

Comment (sureshanaparti) [inline vmware-base/src/com/cloud/hypervisor/vmware/mo/DistributedVirtualSwitchMO.java]: Thanks for the confirmation

**Comment by sureshanaparti** [comment]
Code changes LGTM

**Comment by priyankparihar** [comment]
Hi,
Code is tested on master branch. VR, SSVM, CPVM and  User VM all are coming up properly.
"StartCommand failed due to Exception: com.vmware.vim25.AlreadyExists."  is  not observed during testing of this patch. 


Test procedure followed 
---
Breakpoint is placed before "Line 55: synchronized (uniquedvPortGroupPerDvs.intern())..".
Breakpoint is disabled when multiple threads arrives at barrier(Breakpoint). 
 
This approach is followed, because earlier this issue was coming because of concurrent attempts to create same dv portgroup on same dvSwitch by manager threads of CPVM and SSVM when both are started at same time as mention in issue.

vSphere image PFA. 




img

I would like to give Test LGTM.

**Comment by sateesh-chodapuneedi** [comment]
ping @karuturi

**Comment by sateesh-chodapuneedi** [comment]
tag:mergeready

**Pull request merged**